test/distributed/tensor/parallel/test_shard_embedding.py

```diff
@@ -0,0 +1,109 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+# Owner(s): ["oncall: distributed"]
+
+from copy import deepcopy
+
+import torch
+import torch.nn as nn
+from torch.distributed._tensor import DeviceMesh, Replicate, Shard
+from torch.distributed.tensor.parallel import ColwiseParallel, parallelize_module
+from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
+    parametrize,
+    run_tests,
+)
+from torch.testing._internal.distributed._tensor.common_dtensor import (
+    DTensorTestBase,
+    NUM_DEVICES,
+    skip_unless_torch_gpu,
+    with_comms,
+)
+
+
+class DistShardEmbeddingTest(DTensorTestBase):
+    # Test that sharding behavior is consistent regardless of the order of the sharded
+    # dimension. This would be helpful when we shard multiple embedding modules in sequence
+    # parallelism, where we might need to make element-wise addition (after broadcasting).
+    @with_comms
+    @skip_unless_torch_gpu
+    def test_sharding_consistency_under_reordered_dimension(self):
+        vocab_size, batch_size, seq_len, dim = 25, 5, 8, 13
+        assert seq_len % self.world_size == 0  # output is sharded on seq_len
+
+        torch.manual_seed(5)
+        emb_output_shard_0 = nn.Embedding(vocab_size, dim)
+        emb_output_shard_1 = deepcopy(emb_output_shard_0)
+
+        device_mesh = DeviceMesh(self.device_type, torch.arange(0, NUM_DEVICES))
+        parallelize_module(
+            emb_output_shard_0, device_mesh, ColwiseParallel(output_layouts=Shard(0))
+        )
+        parallelize_module(
+            emb_output_shard_1, device_mesh, ColwiseParallel(output_layouts=Shard(1))
+        )
```

Comment:
For embedding, Colwise means that the weight is always sharded by the last dimension. Are you also trying to test the case when there is a resharding of output?

Reply:
Yes, here I'm trying to test that given a two dimensional input (of size `[batch_size, seq_len]`) to the embedding module, sharding on the `seq_len` dimension give the same result, regardless if it is on `Shard(0)` or `Shard(1)`. This is a simple test on sharding semantics; the test below is an E2E test where we element-wise add `tok_embeddings` (output on `Shard(1)`) and `pos_embeddings` (output on `Shard(0)`).